### PR TITLE
[SDPA/memeff] Backport changes from xFormers to PT

### PR DESCRIPTION
**Pull request by danthe3rd**
Backporting a few fixes from xFormers:
* Bug fixes for local attention (which is not exposed in PT at the moment)
* Massively reduced memory usage on the BW pass (see also https://github.com/facebookresearch/xformers/pull/1028)

Essentially this will also make xFormers build process much easier, as we will be able to use mem-eff from PyTorch (if the user has a recent enough version) rather than building it at xFormers install time
The goal is to have the source of truth for these files in PT moving forward, and remove them from xFormers eventually once our users have a recent-enough version of PT.

### CODE REVIEW

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu [hunk 1]
```diff
@@ -310,9 +310,23 @@ _efficient_attention_backward(
...
+    // a `torch.cat` call in the backward pass
+    at::Tensor chunk = at::empty({B, M, 3, nH, K}, query.options());
+    grad_q = chunk.select(2, 0);
+    grad_k = chunk.select(2, 1);
+    grad_v = chunk.select(2, 2);
+  } else {
+    grad_q = at::empty(query.sizes(), query.options());
+    grad_k = at::empty(key.sizes(), key.options());
+    grad_v = at::empty(value.sizes(), value.options());
+  }
```
Comment (danthe3rd) [inline aten/src/ATen/native/transformers/cuda/attention_backward.cu:329]: TODO: Leave this off by default

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 2]
```diff
@@ -1219,8 +1212,13 @@ std::tuple<Tensor, Tensor, Tensor, Tensor, c10::SymInt, c10::SymInt> _efficient_
           " kb)");
       AT_CUDA_CHECK(err);
     }
+    auto blocks = p.getBlocksGrid();
```
[the remark this answers is not in the document]
Comment (danthe3rd) [reply aten/src/ATen/native/transformers/cuda/attention.cu:1215]: This corresponds to having seqlen for keys/values=0. We still need to zero-out the output in that case

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu [hunk 3]
```diff
@@ -439,8 +464,7 @@ _efficient_attention_backward(
     ASSIGN_CHECK_OVERFLOW(p.gQ_strideH, grad_q.stride(2));
     ASSIGN_CHECK_OVERFLOW(p.gK_strideH, grad_k.stride(2));
     ASSIGN_CHECK_OVERFLOW(p.gV_strideH, grad_v.stride(2));
-    // We removed the chunk/cat optimization and the multiplier is always 1
-    p.gQKV_strideM_multiplier = 1;
+    p.gQKV_strideM_multiplier = grad_q.is_contiguous() ? 1 : 3;
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention_backward.cu:467]: Can you change the ternary to check the is_shared arg

File: aten/src/ATen/native/transformers/cuda/mem_eff_attention/kernel_backward.h [hunk 4]
```diff
@@ -1466,13 +1480,7 @@ struct AttentionBackwardKernel {
...
-      if (p.custom_mask_type == CausalFromTopLeft &&
-          key_start + num_keys_in_block <=
-              int32_t(query_start) - p.window_size) {
-        return;
-      }
-    }
+
     auto prologueGradV = [&](int col) {
       typename MatmulGradV::Mma::IteratorB iterator_dO(
           {int32_t(p.gO_strideM)},
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/mem_eff_attention/kernel_backward.h:1486]: Looks good !

**Review by drisspg — APPROVED** [review]

**Comment by danthe3rd** [comment]
@pytorch-bot merge

**Pull request closed without merging**